### PR TITLE
[AERIE-1648] Extract MissionModelWrapper and redistribute its functionalities

### PR DESCRIPTION
**Pull request by adrienmaillard**
* **Tickets addressed:** AERIE-1648
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

Motivation: many components of the `MissionModelWrapper` class are more appropriately encapsulated by the other existing classes and data structures.

Its functionalities have been mostly redistributed into `Problem`. 

There are two commits: the first is the refactoring, the second one contains the consequences of a linting pass on the scheduler code.

## Verification

Validated against `MerlinsightRules`. 

## Documentation

None

## Future work

None

### CODE REVIEW

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java
Comment (goetzrrGit) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java:117]: Quick question:  `problem` is confusing to me. Are problems errors found in the plan or the constraints/rules the plan needs to follow?

The documentation: ` @param problem the problem that the plan adheres to`

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (goetzrrGit) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: Might want to move the variable declaration to the beginning of the class

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (goetzrrGit) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: Same here, move and group them with the other variables at the top.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java
Comment (goetzrrGit) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java:155]: This is always returning `true`, is that the desired behavior?

**Review by goetzrrGit — CHANGES_REQUESTED** [review]
I just have some small changes and a quick question

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java:155]: Yes, we had implemented something for authenticating through GraphQL but I guess the need has disappeared, (hasn't it?). I am not sure about how authentication would be handled between a scheduler-server and a merlin-server.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java
Comment (adrienmaillard) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java:117]: Maybe this sentence is wrong, that's what you mean? A plan adheres to a mission model but not to a problem. A problem might contain global constraints that need to be enforced during scheduling. And rules/goals that scheduling must try to achieve.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java
Comment (kroffo) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java:479]: Should there still be a check on if the activity type is valid, throwing the `IllegalArgumentException` when it isn't? As written, if the type isn't known, this will just cause a runtime exception.

**Review by kroffo — COMMENTED** [review]
After a single lookover, what I understand makes sense to me, but since much of this is unfamiliar to me I'm going to hold off on an approval until I can take more time to look in more detail at the parts I'm less familiar with.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java:479]: You're right, I've added your suggestion.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: Moved all at the top.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: I have moved them

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java
Comment (goetzrrGit) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java:117]: Oh, I get it now. `Problems` in this context are the thing the scheduler is trying to solve like a math problem. I was reading it as errors or 'here are problems/error with the plan'

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java
Comment (goetzrrGit) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java:155]: Is this a ToDo for later? I don't want to hold up the PR?

Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java:155]: I asked on Slack and it seems that it will be deferred until 0.11.2.

Comment (goetzrrGit) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java:155]: Sounds good. Sorry I missed that slack conversation

**Review by goetzrrGit — APPROVED** [review]

**Review by kroffo — APPROVED** [review]

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java]: :small_red_triangle_down: The message in this exception has duplicate text and is unclear. Where, has the JSON object not been found? Where was the user/system expecting this code to be looking for the activity type?

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java]: :small_red_triangle_down: This TODO is now redundant.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java]: :small_red_triangle_down: This to do is now redundant.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: :large_orange_diamond: Extra new line

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: :large_orange_diamond: To what visualization does this refer to?

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: :large_orange_diamond: Can you fill me in about why this activity type is added/needed.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: :large_orange_diamond: Can you fill me in about what function is performed by these built-in activity types?

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: :large_orange_diamond: Can we talk about this? And generally for any time we might want to add a TODO, I'm of the opinion that we should discuss any TODOs added to the code base to determine if: 
1. they more appropriately deserve a ticket in the back log. 
2. if they don't deserve a ticket, then the extent of the TODO should be confined to the few lines of code around the location of the TODO.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: :large_orange_diamond: Ticket or local change?

**Review by mattdailis — APPROVED** [review]
Looks great!

**Review by patkenneally — CHANGES_REQUESTED** [review]
My review is in progress but posting these questions here now so that it doesn't get merged before we can address them.

:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/NAryMutexConstraint.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/NAryMutexConstraint.java]: 🔻 Remove these fully qualified type names in favor of an import of the type/class.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/OptimizerEarliestEndTime.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/OptimizerEarliestEndTime.java:14]: Can you help me understand how this change is classified as linting/style clean up? This seems to clarify existing expectations or add new behavior. If so, then changes such as these belong in a separate commit so your reviewer is easily able to pick out code clean up from behavior changes and your commit message accurately describes the changes you've made.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: The visualization that would show the valid windows to the user. We do not have it yet.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/TextWriter.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/TextWriter.java]: :large_orange_diamond: Are we using/plan to use this TextWriter class for anything? If not I vote that it be removed.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: They are special types of activities to denote "valid" windows. We used that to track the refinement of windows but I guess we could just remove this now as this is not grounded conceptually yet.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: Yes we can discuss. Basically, we wanted to track valid windows to show them to the user. A better way would be to have a specific data structure for that. I think this is something we will have to do in the future because these kind of windows are useful for the user to debug. It is an intermediate step between a log and a explaination.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/ActivityInstance.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/ActivityInstance.java:166]: :large_blue_circle: My preference would be that such changes are contained in a separate commit, given that I think this goes further than linting/cleaning. I'm happy and open to be convinced otherwise 😄

**Review by patkenneally — CHANGES_REQUESTED** [review]

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: Following my above comment, I'd say ticket.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (patkenneally) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: I agree. Having half way code/not yet grounded concepts through the code base makes it difficult to understand. One has to be to both discern what is true as of now, and what might be true in the future. It requires to much information external to the code to make sense of the code. I vote for removing this and aspects that fall into this category. If we want to hold on to such speculative code then prototype branches might be a good option.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (patkenneally) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: Okay great. Let's add a ticket for this capability and remove this code.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/OptimizerEarliestEndTime.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/OptimizerEarliestEndTime.java:14]: The linter detected that this could be null or not have an end time. So I protected the rest of the call.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/TextWriter.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/TextWriter.java]: I do not know so we can remove it.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/ActivityInstance.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/ActivityInstance.java:166]: It is the direct result of the linting telling that this is a problem.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/OptimizerEarliestEndTime.java
Comment (patkenneally) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/OptimizerEarliestEndTime.java:14]: That's fair. I think these changes lends themselves nicely to being a single commit. The reviewer can infer that the developer was intentional about the very localized nature of the change, making it fast to review.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/TextWriter.java
Comment (patkenneally) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/TextWriter.java]: Great. Let's remove it.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/ActivityInstance.java
Comment (patkenneally) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/ActivityInstance.java:166]: Thanks for the note. It helps my understanding. 👍

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (patkenneally) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: Ticket has been created. The code and the TODOs can be removed.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java]: I replace by "Activity type found in JSON object after request to merlin server has not been found in types extracted from mission model. Probable inconsistency between mission model used by scheduler server and merlin server.", ok ?

Comment (patkenneally) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/AerieController.java]: Great! Very clear.

**Comment by adrienmaillard** [comment]
I have adressed all your comments, Patrick.

**Review by patkenneally — APPROVED** [review]

**Pull request merged**
